lib/logstash/filters/grok.rb

```diff
@@ -288,21 +288,28 @@ def filter(event)
         end
         #break if done
       end # @patterns.each
-  
+
       if matched
         filter_matched(event)
       else
         # Tag this event if we can't parse it. We can use this later to
         # reparse+reindex logs if we improve the patterns given.
         @tag_on_failure.each do |tag|
-          event["tags"] ||= []
-          event["tags"] << tag unless event["tags"].include?(tag)
+          # do not replace the code below with:
+          #   event["tags"] ||= []
+          #   event["tags"] << tag unless event["tags"].include?(tag)
+          # this assumes implementation specific feature of returning a mutable object
+          # from a field ref which should not be assumed and will change in the future.
+
+          tags = event["tags"] || []
```

Comment:
can we just replace this with `event.tag(tag)` and change the [default implementation of tag](https://github.com/elastic/logstash/blob/master/logstash-core-event/lib/logstash/event.rb#L201-L205) to do this?

I think this cleanup fits the whole refactoring you're doing :)